src/raftstore/store/store.rs

```diff
@@ -1536,104 +1536,99 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+            Some(peer) => {
+                peer.mut_store().region = latest.clone();
+                peer.post_split();
+                if peer.is_leader() {
+                    peer.heartbeat_pd(&self.pd_worker);
+                }
+                (peer.peer_stat.clone(), peer.is_leader())
+            }
         };
-
-        {
-            let peer = self.region_peers.get_mut(&region_id).unwrap();
-            peer.mut_store().region = origin_region;
-            peer.post_split();
-        }
-        let new_region_id = new_region.get_id();
-        if let Some(peer) = self.region_peers.get(&new_region_id) {
-            // If the store received a raft msg with the new region raft group
-            // before splitting, it will creates a uninitialized peer.
-            // We can remove this uninitialized peer directly.
-            if peer.get_store().is_initialized() {
-                panic!("duplicated region {} for split region", new_region_id);
+        if is_leader {
+            // Notify pd immediately to let it update the region meta.
+            if let Err(e) = report_split_pd(&regions, &self.pd_worker) {
+                error!("{} failed to notify pd: {}", self.tag, e);
             }
         }
-
-        let mut campaigned = false;
-        let peer;
-        match Peer::create(self, &new_region) {
-            Err(e) => {
-                // peer information is already written into db, can't recover.
-                // there is probably a bug.
-                panic!("create new split region {:?} err {:?}", new_region, e);
-            }
-            Ok(mut new_peer) => {
-                for peer in new_region.get_peers() {
-                    // Add this peer to cache.
-                    new_peer.insert_peer_cache(peer.clone());
-                }
-                peer = new_peer.peer.clone();
-                if let Some(origin_peer) = self.region_peers.get_mut(&region_id) {
-                    // New peer derive write flow from parent region,
-                    // this will be used by balance write flow.
-                    new_peer.peer_stat = origin_peer.peer_stat.clone();
-
-                    campaigned =
-                        new_peer.maybe_campaign(origin_peer, &mut self.pending_raft_groups);
-
-                    if origin_peer.is_leader() {
-                        // Notify pd immediately to let it update the region meta.
-                        if let Err(e) = if right_derive {
-                            report_split_pd(&mut new_peer, origin_peer, &self.pd_worker)
-                        } else {
-                            report_split_pd(origin_peer, &mut new_peer, &self.pd_worker)
-                        } {
-                            error!("{} failed to notify pd: {}", self.tag, e);
-                        }
-                    }
-                }
-
+        let last_key = enc_end_key(regions.last().unwrap());
+        self.region_ranges
+            .remove(&last_key)
```

Comment:
Seems the last_region's `end_key` should be the same with the current region's `end_key`?  Could we use `enc_end_key(latest)` instead here? It seems more clear.

Reply:
I think it depends on whether it is `right_derive`.